sound/soc/sof/ipc.c

```diff
@@ -615,90 +615,75 @@ EXPORT_SYMBOL(snd_sof_ipc_stream_posn);
-		cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
-		cdata->cmd = ctrl_cmd;
-		cdata->type = ctrl_type;
-		cdata->rhdr.hdr.size = scontrol->size;
-		cdata->comp_id = scontrol->comp_id;
-		cdata->num_elems = scontrol->num_channels;
-
-		/* send IPC to the DSP */
-		err = sof_ipc_tx_message(sdev->ipc,
-					 cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size,
-					 cdata, cdata->rhdr.hdr.size);
-		if (err < 0) {
-			dev_err(sdev->dev, "error: failed to set control %d values\n",
-				cdata->comp_id);
-			return err;
+		/* write/read value header via mmaped region */
+		switch (ctrl_type) {
+		case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+		case SOF_CTRL_TYPE_VALUE_COMP_GET:
+		case SOF_CTRL_TYPE_DATA_GET:
+			send = 0;
+			break;
+		case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		case SOF_CTRL_TYPE_DATA_SET:
+			send = 1;
+			break;
+		default:
+			return -EINVAL;
 		}
-	}
+		send_bytes = sizeof(struct sof_ipc_ctrl_value_chan) *
+		cdata->num_elems;
 
-	return 0;
-}
-EXPORT_SYMBOL(snd_sof_ipc_set_comp_data);
+		if (send)
+			snd_sof_dsp_block_write(sdev, sdev->mmio_bar,
+						scontrol->readback_offset,
+						cdata->chanv, send_bytes);
 
-int snd_sof_ipc_get_comp_data(struct snd_sof_ipc *ipc,
-			      struct snd_sof_control *scontrol, u32 ipc_cmd,
-			      enum sof_ipc_ctrl_type ctrl_type,
-			      enum sof_ipc_ctrl_cmd ctrl_cmd)
-{
-	struct snd_sof_dev *sdev = ipc->sdev;
-	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	int err;
-
-	/* read firmware byte counters */
-	if (scontrol->readback_offset != 0) {
-		/* we can read values via mmaped region */
-		snd_sof_dsp_block_read(sdev, sdev->mmio_bar,
-				       scontrol->readback_offset, cdata->chanv,
-				       sizeof(struct sof_ipc_ctrl_value_chan) *
-				       cdata->num_elems);
+		else
+			snd_sof_dsp_block_read(sdev, sdev->mmio_bar,
+					       scontrol->readback_offset,
+					       cdata->chanv, send_bytes);
+		return err;
+	}
 
-	} else {
-		/* read position via slower IPC */
-		cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
-		cdata->cmd = ctrl_cmd;
-		cdata->type = ctrl_type;
-		cdata->rhdr.hdr.size = scontrol->size;
-		cdata->comp_id = scontrol->comp_id;
-		cdata->num_elems = scontrol->num_channels;
-
-		/* send IPC to the DSP */
-		err = sof_ipc_tx_message(sdev->ipc,
-					 cdata->rhdr.hdr.cmd, cdata,
-					 cdata->rhdr.hdr.size,
-					 cdata, cdata->rhdr.hdr.size);
-		if (err < 0) {
-			dev_err(sdev->dev, "error: failed to get control %d values\n",
-				cdata->comp_id);
-			return err;
-		}
+	cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
+	cdata->cmd = ctrl_cmd;
+	cdata->type = ctrl_type;
+	cdata->rhdr.hdr.size = scontrol->size;
+	cdata->comp_id = scontrol->comp_id;
+	cdata->num_elems = scontrol->num_channels;
+
+	/* write value via slower IPC */
+	if (cdata->rhdr.hdr.size > SOF_IPC_MSG_MAX_SIZE) {
+		dev_err(sdev->dev, "error: set/get size %d too big comp %d\n",
+			cdata->rhdr.hdr.size, cdata->comp_id);
+		return err;
```

Comment:
here err is zero, so that's not a error that will be trapped by upper layers.

Reply:
also show the message size check be part of sof_ipc_tx_message?

Reply:
tried to fix. this is a patch which is replaced by the next one, so hope it went ok. I think the ipc_tx message size is eventually checked somewhere. I've tried to send big messages without chopping and they fail in sending with size check.